File: netlify/functions/recommendation.js
import { spawnSync } from 'node:child_process';
import path from 'node:path';

const BACKEND_RUNNER = path.join(__dirname, 'backend', 'run_recommendation.py');

const corsHeaders = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Headers': 'Content-Type',
};

export const handler = async (event) => {
  if (event.httpMethod === 'OPTIONS') {
    return { statusCode: 200, headers: corsHeaders, body: '' };
  }

  let payload;
  try {
    payload = JSON.parse(event.body || '{}');
  } catch (error) {
    return {
      statusCode: 400,
      headers: corsHeaders,
      body: JSON.stringify({ error: 'Invalid JSON payload' }),
    };
  }

  if (!payload.data) {
    return {
      statusCode: 400,
      headers: corsHeaders,
      body: JSON.stringify({ error: 'Playlist link is required' }),
    };
  }

  const pythonResult = spawnSync('python3', [BACKEND_RUNNER], {
    input: JSON.stringify(payload),
    encoding: 'utf-8',
    env: { ...process.env },
  });

  if (pythonResult.error) {
    return {
      statusCode: 500,
      headers: corsHeaders,
      body: JSON.stringify({ error: `Python execution failed: ${pythonResult.error.message}` }),
    };
  }

  if (pythonResult.status !== 0) {
    return {
      statusCode: 500,
      headers: corsHeaders,
      body: JSON.stringify({ error: pythonResult.stderr || 'Backend error' }),
    };
  }

  try {
    const data = JSON.parse(pythonResult.stdout || '{}');
    if (data.error) {
      return { statusCode: 500, headers: corsHeaders, body: JSON.stringify(data) };
    }
    return {
      statusCode: 200,
      headers: { ...corsHeaders, 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    };
  } catch (error) {
    return {
      statusCode: 500,
      headers: corsHeaders,
      body: JSON.stringify({ error: 'Unable to parse backend response' }),
    };
  }
};

File: netlify/functions/recommendation.py
import json
import sys
from pathlib import Path

# Ensure the backend package is importable whether running locally or in Netlify Functions
CURRENT_DIR = Path(__file__).resolve().parent
BACKEND_PATH = CURRENT_DIR / "backend"
sys.path.append(str(BACKEND_PATH.parent))

from backend.recommend import recommender  # noqa: E402


def _cors_headers():
    return {
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Headers": "Content-Type",
    }


def handler(event, context):
    if event.get("httpMethod") == "OPTIONS":
        return {"statusCode": 200, "headers": _cors_headers(), "body": ""}

    try:
        payload = json.loads(event.get("body") or "{}")
    except json.JSONDecodeError:
        return {
            "statusCode": 400,
            "headers": _cors_headers(),
            "body": json.dumps({"error": "Invalid JSON payload"}),
        }

    link = payload.get("data")
    sort_method = payload.get("sort_method", "heap")

    if not link:
        return {
            "statusCode": 400,
            "headers": _cors_headers(),
            "body": json.dumps({"error": "Playlist link is required"}),
        }

    try:
        recommendation = recommender(link, sort_method)
        result = recommendation.get_recommendations()
    except Exception as exc:  # pragma: no cover - defensive guard for lambda runtime
        return {
            "statusCode": 500,
            "headers": _cors_headers(),
            "body": json.dumps({"error": str(exc)}),
        }

    return {
        "statusCode": 200,
        "headers": {**_cors_headers(), "Content-Type": "application/json"},
        "body": json.dumps({"recommendations": result}),
    }
